core/src/fixed_window.rs

```diff
@@ -0,0 +1,206 @@
+use std::borrow::Borrow;
+use std::cmp::Eq;
+use std::collections::hash_map::Entry::{Occupied, Vacant};
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::ops::AddAssign;
+use std::time::{Duration, Instant};
+
+/// A fixed lookback window maintains some statistics over the half-open time
+/// interval:
+///    [start  <- duration ->  end)
+/// When time passes the end of the window, a new window is instantiated.
+/// Compared to "rolling" windows, it has the advantage of computational and
+/// memory efficiency. However, at the beginning of each window, it starts from
+/// scratch and has zero data. To mitigate this, this implementation uses two
+/// windows simultaneously, one starting half the interval after the other.
+/// When looking up the values, it uses the interval which starts earlier.
+pub struct FixedWindow<K, V> {
+    /// Length of lookback window.
+    duration: Duration,
+    /// Beginning of each lookback window.
+    /// One should start half the duration after the other.
+    epoch: [Instant; 2],
```

Comment:
Nit: `epoch` in the system today implies a very specific range of `blocks`, (looking at `DEFAULT_SLOTS_PER_EPOCH`: https://github.com/solana-labs/solana/blob/master/sdk/src/clock.rs#L23.

 Maybe rename this to `window_start_ts`, for  window start timestamp?